src/Tests/Tests/ClientConcepts/HighLevel/Serialization/CustomSerialization.doc.cs

```diff
@@ -242,5 +249,112 @@ public class MyDocument
 			// hide
 			Expect(expected, preserveNullInExpected: true).FromRequest(indexResponse);
 		}
+
+		/** ==== Serializing Type Information
+		 * Here's another example that implements a custom contract resolver. The custom contract resolver
+		 * will include the type name within the serialized JSON for the document, which can be useful when
+		 * returning covariant document types within a collection.
+		 */
+		public class MySecondCustomContractResolver : ConnectionSettingsAwareContractResolver
+		{
+			public MySecondCustomContractResolver(IConnectionSettingsValues connectionSettings) : base(connectionSettings)
+			{
+			}
+
+			protected override JsonContract CreateContract(Type objectType)
+			{
+				var contract = base.CreateContract(objectType);
+				if (contract is JsonContainerContract containerContract)
+				{
+					if (containerContract.ItemTypeNameHandling == null)
+						containerContract.ItemTypeNameHandling = TypeNameHandling.None;
+				}
+
+				return contract;
+			}
+		}
+
+		public class MySecondCustomJsonNetSerializer : ConnectionSettingsAwareSerializerBase
+		{
+			public MySecondCustomJsonNetSerializer(IElasticsearchSerializer builtinSerializer, IConnectionSettingsValues connectionSettings)
+				: base(builtinSerializer, connectionSettings) { }
+
+			protected override JsonSerializerSettings CreateJsonSerializerSettings() =>
+				new JsonSerializerSettings
+				{
+					TypeNameHandling = TypeNameHandling.All,
+					NullValueHandling = NullValueHandling.Ignore,
+					TypeNameAssemblyFormatHandling = TypeNameAssemblyFormatHandling.Simple
+				};
+
+			protected override ConnectionSettingsAwareContractResolver CreateContractResolver() =>
+				new MySecondCustomContractResolver(ConnectionSettings); // <1> override the contract resolver
+		}
+
+		/**
+		 * Now, hooking up this serializer
+         */
+		[U] public void MySecondJsonNetSerializer()
+		{
+			var pool = new SingleNodeConnectionPool(new Uri("http://localhost:9200"));
+			var connectionSettings = new ConnectionSettings(
+					pool,
+					connection: new InMemoryConnection(),
+					sourceSerializer: (builtin, settings) => new MySecondCustomJsonNetSerializer(builtin, settings))
+				.DefaultIndex("my-index");
+
+			//hide
+			connectionSettings.DisableDirectStreaming();
+
+			var client = new ElasticClient(connectionSettings);
+
+			/** and indexing an instance of our document type */
+			var document = new MyDocument
+			{
+				Id = 1,
+				Name = "My first document",
+				OwnerId = 2,
+				SubDocuments = new []
+				{
+					new MySubDocument { Name = "my first sub document" },
+					new MySubDocument { Name = "my second sub document" },
+				}
+			};
+
+			var indexResponse = client.IndexDocument(document);
+
+			/** serializes to */
+			//json
+			var expected = new JObject
+			{
+				{ "$type", "Tests.ClientConcepts.HighLevel.Serialization.GettingStarted+MyDocument, Tests" },
```

Comment:
`nameof`?

Reply:
`nameof` only resolves to the leaf value sadly. Could use `AssemblyQualifiedName` and strip everything after assembly name (seems easiest way)

Reply:
Will keep as string because the `JObject` is serialized to JSON in documentation and other approaches complicate this.